V5 api design

SimonBlanke opened the pull request:
@fkiraly I just moved most of your code in my optimizer.py-file. I added multiple examples for different kinds of experiments. The provided experiments should show how the user would create his own experiments.

fkiraly commented:
Nice!

Question upon superficial review - why are there two sklearn experiments in the code?

SimonBlanke commented:
The experiments I added are not integrations, but just example experiments a user could create on his own when working with hyperactive. So the purpose of these experiments is to show case how users could create their custom experiments with the `BaseExperiment`-class. 

So this is not source-code, but should facilitate a common understanding in how Experiments are created and used.

SimonBlanke commented:
I moved the v4 Code into the "src_old"-directory for now, because it is easier to look up stuff.

The code in the "src"-directory implements a working prototype (without multiprocessing and some other stuff) for the v5 API. You can try it out with the script in the following path "examples/v5_API_example/example.py"

I added some minor changes. In the BaseExperiment I changed the parameters, so that the catch- and callbacks-parameters are passed to init. The parameters for the model (cross-cv, X_train, y_train, ...) can be passed separately via the "setup"-method. This is also shown in the example script. I wanted to separate those user defined parameters from the API parameters. It also feels correct to move catch and callbacks to the experiment, because they affect the objective-function/score-method.

I added a class, that creates the search-space in:
src/hyperactive/search_config/_search_config.py
This can be part of the public API (current implementation) or we can move it to the backend.

There is one problem still. Hyperactive v4 has a feature called OptimizationStrategies, in which you can chain together optimization algorithms. Here is an example:
https://github.com/SimonBlanke/Hyperactive/blob/master/examples/opt_strat_early_stop.py#L14

Here is some documentation:
https://simonblanke.github.io/hyperactive-documentation/4.7/tutorials/optimization_strategies/

I do not know how to build this into the current API. The Optimizer class is currently our "runner", because it has "add_search" and "run". It looks awkward to me to have an OptimizationStrategies API, that would look like this:

```python
opt_strat = CustomOptimizationStrategy()

opt_strat.add_optimizer(RandomSearchOptimizer())
opt_strat.add_optimizer(HillClimbingOptimizer())

opt_strat.add_search(...)
opt_strat.run()
```
The mixture of "add_optimizer" and "add_search" in one class does not seem right to me.

fkiraly commented:
May I request more information about `CustomOptimizationStrategy` - what is the mathematical, or algorithmic object that this class represents?

SimonBlanke commented:
> May I request more information about CustomOptimizationStrategy - what is the exact mathematical, or algorithmic object that this class represents?

This concept was just an idea of mine and not rooted in a specific mathematical or algorithmic object. I saw, that some algorithms (like hill-climbing) have some obvious drawbacks, that could be compensated by combining them with other algorithms (like random-search). So exploring the search-space first with a random-search and then passing the best-positions from it to the hill-climbing search leads to good results (from my experience) in general.
Sometimes users don't want to dig through many optimization algorithms to find the best one for their problem. Of course the user could just manually run a random-search and pass the best positions and other data to a hill-climbing search, which "continues" the search. But the OptimizationStrategy-concept aims to provide a better "out of the box" experience.

The syntax you provides looks intriguing. I am also open to the idea to remove the support for this optimization-strategy-concept for v5. We can later re-integrate it properly (with your suggestions). This would speed up the migration to v5.

fkiraly commented:
>  saw, that some algorithms (like hill-climbing) have some obvious drawbacks, that could be compensated by combining them with other algorithms (like random-search). So exploring the search-space first with a random-search and then passing the best-positions from it to the hill-climbing search leads to good results (from my experience) in general.

Ok, so I believe I understood the algorithm correctly.

In that case I would be very certain that the proposed syntax is the "typical `sklearn`-like API" for the kind of meta-algorithm proposed.

It is a special case of the composite/strategy combination pattern used frequently there, see `sklearn` `Pipeline` - what you are doing is basically a pipeline that passes on the optimal value to the next element.

SimonBlanke commented:
I agree with your assessment of the optimization-strategy feature. It is very similar to an scikit-learn pipeline. I opened a separate issue for this discussion: [SimonBlanke/Hyperactive/issues/103](https://github.com/SimonBlanke/Hyperactive/issues/103)

The last major feature to be adapted to v5 is the multiprocessing. I would like to finish this first. We should also discuss how a CompositeOptimizer would do multiprocessing from an API perspective.

fkiraly commented:
can you point me to an example vignette or tutorial for multiprocessing?

One question I have where it comes into play. A pipeline is sequential, so the steps cannot be parallelized, at least not in general or without adding an algorithm.

SimonBlanke commented:
> can you point me to an example vignette or tutorial for multiprocessing?

Hyperactive works with a multiprocessing pool. This is explained in the first few chapters [of this page](https://superfastpython.com/multiprocessing-pool-python/#What_Is_a_Process_Pool).

> A pipeline is sequential, so the steps cannot be parallelized, at least not in general or without adding an algorithm.

Parallelizing each step is not necessary here. The multiprocessing would run each pipeline separately (except for a [shared memory-dictionary](https://github.com/SimonBlanke/Hyperactive/blob/master/src/hyperactive/hyperactive.py#L53), that I will take care of). I am just curious how this would look like from an API perspective. I guess it would look similar to the regular optimizer API:

```python
pipeline1 = OptimizerPipeline1(optimizer_list1)
pipeline2 = OptimizerPipeline2(optimizer_list2)

pipeline1.add_search(experiment1)
pipeline2.add_search(experiment2)

hyper = pipeline1 + pipeline2
hyper.run()
```

SimonBlanke commented:
Multiprocessing should work now. It is also possible to add more than two optimizers like this:
```python
hyper = optimizer1 + optimizer2 + optimizer3 + optimizer4
```

I will keep looking into the tests and examples to see if further fixes are necessary.

fkiraly commented:
Nice!

FYI, in `sktime`, we have `*` for sequential, and `+` for parallel.

fkiraly commented on src/hyperactive/base/_experiment.py:
```diff
@@ -0,0 +1,79 @@
+    ):
+        super().__init__()
+
+        self.callbacks = callbacks or {}
```
why is this in the base class? The args are not used.

Also, if we want these, we need to go with the dataclass-like API of sklearn, the parameters should never change from their value as init arg.

SimonBlanke replied:
The catch and callbacks parameters are used in [the objective-function source code](https://github.com/SimonBlanke/Hyperactive/blob/v5-API-design/src/hyperactive/optimizers/objective_function.py#L23-L44). I removed those two parameters from the `add_search`-method, because they are used in the objective-funtion, which is part of the experiment.

> Also, if we want these, we need to go with the dataclass-like API of sklearn, the parameters should never change from their value as init arg.

The `self.callbacks = callbacks or {}` is used to avoid `{}` as a default argument.

SimonBlanke commented:
I just changed the API for the catch and callback parameters. They now work as decorators of the objective function. The implementation can be seen in the [experiment utility](https://github.com/SimonBlanke/Hyperactive/pull/101/files#diff-05e18ef408f33dda796d4d9658978c587ebac11ac8fe05bec1dd87775a702687).

And it looks like [this](https://github.com/SimonBlanke/Hyperactive/pull/101/files#diff-a40efe9cbb5906753295077398dd70a79c7ff3e828b72add0d188afd698e6584) in action.

There are multiple advantages:
 - the callback-functions can be moved to the experiment class (which is already nice). This also allows `self` in the function. This way the user can access the variables from the `setup`-method.
 - The implementation for those two parameters is moved out, which decreases the complexity of `search.py`.
 - These parameters works very well with decorators. They already wrapped the objective-function before, but now it is separate from the rest of the source-code.
